Fix for issue: 12050

devcorrelator opened the pull request:
Resolves https://github.com/cakephp/cakephp/issues/12050

devcorrelator committed:
Merge pull request #1 from cakephp/master

merging from the main fork.

devcorrelator committed:
Add checking for prefix setting.

devcorrelator committed:
Added checking for the url parts.

devcorrelator committed:
Derp. forgot to close function.

dereuromark commented on src/Routing/RouteBuilder.php:
```diff
@@ -747,28 +747,27 @@ protected static function parseDefaults($defaults)
             unset($matches[0]);
             $matches = array_filter($matches, function ($value) {
-                return $value !== '' && $value !== '::';
+                return $value !== '::';
```
Why removing empty string check here? That was on purpose afaik.

devcorrelator replied:
That's what was removing $matches[2]. With it there, it could mess with the switch as it relies on the count of the array verse specifically looking to see if like $matches[4] is set.

devcorrelator replied:
Plus, the switch is referencing specific indexes within the $marches array, and only returning specific indexes, so if $matches[11] existed because this was taken it out, ultimately it wouldn't matter (other then maybe garbage collection if $matches is huge for some reason).

devcorrelator replied:
Also, I guess it should return an empty string instead of null? 
See failed tests:
https://travis-ci.org/cakephp/cakephp/jobs/375121707

dereuromark replied:
This is a filter method on the array, any empty value is being stripped, same as :: which is similar empty value.

markstory commented:
These changes cause several tests to fail. We'll need to find another solution that doesn't break existing behavior.

markstory committed:
Simplify route target parsing.

Using the number of elements doesn't give enough context to determine
whether the route points at a plugin or a prefix. Using named capture
groups lets us handle that situation.

Refs #12051

markstory commented:
I've pushed up a change that solves the issue you reported and solves failing tests.
